File: src/services/note.service.js
import { Note } from "../models/note.model.js";
import { NoteStyle } from "../models/noteStyle.model.js";
import { StringError } from "../errors/string.error.js";

const lookupNoteStyleByName = async (styleName) => {
    try {
        const noteStyle = await NoteStyle.findOne({ style_name: styleName });
        if (!noteStyle) {
            throw new StringError(`Note style with name '${styleName}' not found`);
        }
        return noteStyle._id;
    } catch (error) {
        if (error instanceof StringError) {
            throw error;
        }
        console.error("Error looking up note style:", error);
        throw new StringError("Failed to lookup note style");
    }
}

const createNote = async (noteData) => {
    try {
        let finalNoteData = { ...noteData };
        
        // If note_style is a string (name) instead of ObjectId, look up the ID
        if (typeof noteData.note_style === 'string' && !noteData.note_style.match(/^[0-9a-fA-F]{24}$/)) {
            finalNoteData.note_style = await lookupNoteStyleByName(noteData.note_style);
        }

        const note = await Note.create(finalNoteData);
        return note;
    } catch (error) {
        console.error("Error creating note:", error);
        if (error instanceof StringError) {
            throw error;
        }
        throw new StringError("Failed to create note");
    }
}

const getNotes = async () => {
    try {
        const notes = await Note.find();
        return notes;
    } catch (error) {
        console.error("Error fetching notes:", error);
        throw new Error("Failed to fetch notes");
    }
}

const updateNote = async (noteId, updatedData) => {
    try {
        const note = await Note.findByIdAndUpdate(noteId, updatedData, { new: true });
        return note;
    } catch (error) {
        console.error("Error updating note:", error);
        throw new Error("Failed to update note");
    }
}

const deleteNote = async (noteId) => {
    try {
        await Note.findByIdAndDelete(noteId);
    } catch (error) {
        console.error("Error deleting note:", error);
        throw new Error("Failed to delete note");
    }
}

const saveAudioNote = async (req, res) => {
    try {
        const { audio_note } = req.body;
        if (!audio_note) {
            throw new StringError("Audio note is required");
        }
        const noteData = { audio_note };
        const note = await createNote(noteData);
        return res.status(201).json({ status: 1, message: "Audio note saved successfully", data: note });
    } catch (error) {
        console.error("Error saving audio note:", error);
        return res.status(400).json({ status: 0, message: error.message });
    }
}

export default { createNote, getNotes, updateNote, deleteNote, saveAudioNote };

File: flutter_app/lib/services/note_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../configs/network_config.dart';
import '../constants/api_endpoints.dart';
import '../models/note.dart';

class NoteService {
  final Dio _dio = NetworkConfig.dio;

  // Create a new text note
  Future<Map<String, dynamic>> createNote({
    required String noteTitle,
    required String noteStyle,
    String textNote = '',
    String audioTranscription = '',
    String aiNote = '',
  }) async {
    try {
      debugPrint('📝 NoteService: Creating note - $noteTitle');

      final response = await _dio.post(
        ApiEndpoints.userCreateNote,
        data: {
          'note_title': noteTitle,
          'note_style': noteStyle,
          'text_note': textNote,
          'audio_transcription': audioTranscription,
          'ai_note': aiNote,
        },
      );

      debugPrint('📝 NoteService: Note created successfully');
      return response.data;
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Create note error - ${e.message}');
      rethrow;
    }
  }

  // Get all notes
  Future<List<Note>> getNotes() async {
    try {
      debugPrint('📝 NoteService: Fetching all notes');

      final response = await _dio.get(ApiEndpoints.userGetNotes);

      if (response.data['STATUS'] == 1 && response.data['RESULT'] != null) {
        final List<dynamic> notesJson = response.data['RESULT'];
        final notes = notesJson
            .map((json) => Note.fromBackendJson(json))
            .toList();
        debugPrint('📝 NoteService: Fetched ${notes.length} notes');
        return notes;
      }

      return [];
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Get notes error - ${e.message}');
      rethrow;
    }
  }

  // Update an existing note
  Future<Map<String, dynamic>> updateNote({
    required String noteId,
    String? noteTitle,
    String? noteStyle,
    String? textNote,
    String? audioTranscription,
    String? aiNote,
  }) async {
    try {
      debugPrint('📝 NoteService: Updating note - $noteId');

      final Map<String, dynamic> data = {};
      if (noteTitle != null) data['note_title'] = noteTitle;
      if (noteStyle != null) data['note_style'] = noteStyle;
      if (textNote != null) data['text_note'] = textNote;
      if (audioTranscription != null) {
        data['audio_transcription'] = audioTranscription;
      }
      if (aiNote != null) data['ai_note'] = aiNote;

      final response = await _dio.put(
        '${ApiEndpoints.userUpdateNote}/$noteId',
        data: data,
      );

      debugPrint('📝 NoteService: Note updated successfully');
      return response.data;
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Update note error - ${e.message}');
      rethrow;
    }
  }

  // Delete a note
  Future<void> deleteNote(String noteId) async {
    try {
      debugPrint('📝 NoteService: Deleting note - $noteId');

      await _dio.delete('${ApiEndpoints.userDeleteNote}/$noteId');

      debugPrint('📝 NoteService: Note deleted successfully');
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Delete note error - ${e.message}');
      rethrow;
    }
  }

  // Save audio note (upload audio to cloudinary and create note)
  Future<Map<String, dynamic>> saveAudioNote({
    required File audioFile,
    required String noteTitle,
    required String noteStyle,
    String textNote = '',
    String audioTranscription = '',
    String aiNote = '',
  }) async {
    try {
      debugPrint('🎤 NoteService: Saving audio note - $noteTitle');
      debugPrint('🎤 NoteService: Audio file path - ${audioFile.path}');

      final formData = FormData.fromMap({
        'audio': await MultipartFile.fromFile(
          audioFile.path,
          filename: audioFile.path.split('/').last,
        ),
        'note_title': noteTitle,
        'note_style': noteStyle,
        'text_note': textNote,
        'audio_transcription': audioTranscription,
        'ai_note': aiNote,
      });

      final response = await _dio.post(
        ApiEndpoints.userSaveAudioNote,
        data: formData,
        options: Options(headers: {'Content-Type': 'multipart/form-data'}),
      );

      debugPrint('🎤 NoteService: Audio note saved successfully');
      return response.data;
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Save audio note error - ${e.message}');
      rethrow;
    }
  }

  // Process audio note (transcribe and generate AI content)
  Future<Map<String, dynamic>> processAudioNote({
    required File audioFile,
    required String styleName,
  }) async {
    try {
      debugPrint(
        '🤖 NoteService: Processing audio note with style - $styleName',
      );
      debugPrint('🤖 NoteService: Audio file path - ${audioFile.path}');

      final formData = FormData.fromMap({
        'audioFile': await MultipartFile.fromFile(
          audioFile.path,
          filename: audioFile.path.split('/').last,
        ),
        'styleName': styleName,
      });

      final response = await _dio.post(
        ApiEndpoints.userProcessAudioNote,
        data: formData,
        options: Options(headers: {'Content-Type': 'multipart/form-data'}),
      );

      debugPrint('🤖 NoteService: Audio note processed successfully');
      debugPrint('🤖 NoteService: Response - ${response.data}');
      return response.data;
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Process audio note error - ${e.message}');
      rethrow;
    }
  }

  // Get available note styles
  Future<List<Map<String, dynamic>>> getNoteStyles() async {
    try {
      debugPrint('🎨 NoteService: Fetching note styles');

      final response = await _dio.get(ApiEndpoints.userGetNoteStyles);

      if (response.data['STATUS'] == 1 && response.data['RESULT'] != null) {
        final List<dynamic> stylesJson = response.data['RESULT'];
        debugPrint('🎨 NoteService: Fetched ${stylesJson.length} styles');
        return stylesJson.map((e) => Map<String, dynamic>.from(e)).toList();
      }

      return [];
    } on DioException catch (e) {
      debugPrint('❌ NoteService: Get note styles error - ${e.message}');
      rethrow;
    }
  }
}
